torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <c10/macros/Export.h>
+
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//! Traverse and collect all concretized broadcast domains.
+//!
+//! The traversal first initializes the origin map with broadcast
+//! domains in input tensors. Then, a new entry is added to the origin
+//! map when a broadcast op is encountered during a forward traversal
+//! of the given fusion. For non-broadcast ops, mappings are just
+//! propagated forward using PairwiseRootDomainMap.
+//!
+//! When the mapped consumer domain is not broadcast, it means the
+//! producer broadcast domain is concretized, and its origin broadcast
+//! domains are marked as concretized.
+class TORCH_CUDA_CU_API ConcretizedBroadcastDomains : private IterVisitor {
+ public:
+  void build(Fusion* fusion);
+
+  bool isConcretized(IterDomain* id) const;
+
+ private:
+  using IterVisitor::handle;
+
+  void handle(BroadcastOp* bop) final;
+
+  void handle(Expr* expr) final;
+
+  void markAsConcretized(IterDomain* root_domain);
+
+ private:
+  //! Maps each broadcast domain to its original broadcast
+  //! domains. Their can be multiple original domains due to, e.g.,
+  //! binary ops with broadcast domains in both inputs.
+  std::unordered_map<IterDomain*, std::unordered_set<IterDomain*>>
+      broadcast_origin_map_;
+  //! Set of all concretized original domains
+  std::unordered_set<IterDomain*> concretized_domains_;
+};
+
+//! A broadcast expression is trivial when it can be translated to
+//! just a unary Set op. This class gathers all such broadcast
+//! expressions.
+//!
+//! Broadcast is trivial when either of the below is true:
+//! - Not parallelized
+//! - Not thread-predicated
+//! - Parallelized with TID but broadcasted into SMEM
+//! - Not concretized
+//!
+//! TODO: consider other potentially trivial cases. E.g., when a
+//! producer is on SMEM, even if its predicated and parallelized with
+//! TID, all threads can just copy the producer to the consumer,
+//! assuming a proper RAW sync is inserted.
+class TORCH_CUDA_CU_API TrivialBroadcastInfo {
+ public:
+  void build(Fusion* fusion);
+
+  //! Return true when a broadcast expression is trivial
+  bool isTrivial(BroadcastOp* bop) const;
+
+ private:
+  //! Set of trivial broadcast domains
+  std::unordered_set<BroadcastOp*> trivial_bops_;
+};
```

Comment:
It turned out that this isn't needed. Is there any potential use case? Otherwise, I'll remove it.

Reply:
I think doing what you do in codegen is more convenient than blanket marking broadcast ops as trivial or not. The only thing I might suggest in a follow up is just avoid generating GridBroadcast is it's not broadcasted on the grid. I wonder if we should have explicit blockBroadcast/Reduce nodes, so we can do the same and avoid the complexity later in codegen. If so to be done in a follow up PR, not this one.

Reply:
Removed this code.

Reply:
Naoya mentioned to me `lower_index` checks `getParallelBroadcastDomains` before it generates a grid broadcast, so it won't generate a node anymore if not a real grid broadcast.